libs/platform/ebpf_hash_table.c

```diff
@@ -3,52 +3,344 @@
+ * @param[in] length Length of key to hash.
+ * @param[in] seed Seed to randomize hash.
+ * @return Hash of key.
+ */
+unsigned long
+_ebpf_murmur3_32(_In_ const uint8_t* key, size_t length, uint32_t seed)
 {
-    ebpf_hash_table_t* table = (ebpf_hash_table_t*)avl_table;
+    uint32_t c1 = 0xcc9e2d51;
+    uint32_t c2 = 0x1b873593;
+    uint32_t r1 = 15;
+    uint32_t r2 = 13;
+    uint32_t m = 5;
+    uint32_t n = 0xe6546b64;
+    uint32_t hash = seed;
+
+    for (size_t index = 0; (length - index) > 3; index += 4) {
+        uint32_t k = *(uint32_t*)(key + index);
+        k *= c1;
+        k = _ebpf_rol(k, r1);
+        k *= c2;
+
+        hash ^= k;
+        hash = _ebpf_rol(hash, r2);
+        hash *= m;
+        hash += n;
+    }
+    unsigned long remainder = 0;
+    for (size_t index = length & (~3); index < length; index++) {
+        remainder <<= 8;
+        remainder |= key[index];
+    }
+    remainder *= c1;
+    remainder = _ebpf_rol(remainder, r1);
+    remainder *= c2;
+
+    hash ^= remainder;
+    hash ^= (uint32_t)length;
+    hash *= 0x85ebca6b;
+    hash ^= (hash >> r2);
+    hash *= 0xc2b2ae35;
+    hash ^= (hash >> 16);
+    return hash;
+}
 
-    if (table->compare_function) {
-        return (RTL_GENERIC_COMPARE_RESULTS)table->compare_function(
-            (const uint8_t*)first_struct, (const uint8_t*)second_struct);
+/**
+ * @brief Given two potentially non-comparable key values, extract the key and
+ * compare them.
+ *
+ * @param[in] hash_table Hash table the keys belong to.
+ * @param[in] key_a First key
+ * @param[in] key_b Second key
+ * @retval -1 if key_a < key_b
+ * @retval 0 if key_a == key_b
+ * @retval 1 if key_a > key_b
+ */
+static int
+_ebpf_hash_table_compare(ebpf_hash_table_t* hash_table, _In_ const uint8_t* key_a, _In_ const uint8_t* key_b)
+{
+    size_t length_a;
+    size_t length_b;
+    const uint8_t* data_a;
+    const uint8_t* data_b;
+    if (hash_table->extract) {
+        hash_table->extract(key_a, &data_a, &length_a);
+        hash_table->extract(key_b, &data_b, &length_b);
     } else {
-        int result = memcmp(first_struct, second_struct, table->key_size);
-        if (result < 0) {
-            return GenericLessThan;
-        } else if (result > 0) {
-            return GenericGreaterThan;
-        } else {
-            return GenericEqual;
-        }
+        length_a = hash_table->key_size;
+        data_a = key_a;
+        length_b = hash_table->key_size;
+        data_b = key_b;
+    }
+    if (length_a < length_b) {
+        return -1;
+    }
+    if (length_a > length_b) {
+        return 1;
     }
+    return memcmp(data_a, data_b, length_a);
 }
 
-static void*
-_ebpf_hash_table_allocate(struct _RTL_AVL_TABLE* avl_table, unsigned long byte_size)
+/**
+ * @brief Given a potentially non-comparable key value, extract the key and
+ * compute the hash.
+ *
+ * @param[in] hash_table Hash table the keys belong to.
+ * @param[in] key_a First key
+ * @retval Hash of key
```

Comment:
```suggestion
 * @return Hash of key
```
(only use retval for discrete values)